:seedling: Add KubeVirt support to Tilt dev workflow

johananl opened the pull request:
<!-- Thanks for sending a pull request! Here are some tips for you:
    1. If this is your first time, please read our contributor guidelines: https://github.com/kubernetes-sigs/cluster-api/blob/main/CONTRIBUTING.md#contributing-a-patch and developer guide https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/book/src/developer/getting-started.md

    2. Please add an icon to the title of this PR (see https://sigs.k8s.io/cluster-api/CONTRIBUTING.md#contributing-a-patch), and delete this line and similar ones
    the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) 
-->

**What this PR does / why we need it**:

This PR adds support for local development using KubeVirt as an alternative to CAPD.

This is useful in cases where CAPD can't be used for whatever reason, with one example being developing Ignition-related features (since Ignition runs in early boot and therefore can't be containerized easily).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: None

<!-- 
Please label this pull request according to what area(s) you are addressing. For reference on PR/issue labels, see: https://github.com/kubernetes-sigs/cluster-api/labels?q=area+

Area example:
/area runtime-sdk
-->

/area devtools

sbueringer commented:
Is there someone who is familiar with KubeVirt that could do a first review?

nunnatsa commented:
Hi @johananl 

Thanks for this PR. As far as I can see, this is based on the KubeVirt sections in the quickstart guide.

I just opened a PR to update it, as it didn't woek for me now (I was working when I wrote it...). Maybe you want to take a look?

https://github.com/kubernetes-sigs/cluster-api/pull/11734

fabriziopandini commented on docs/book/src/developer/core/tilt.md:
```diff
@@ -8,26 +8,154 @@ workflow that offers easy deployments and rapid iterative builds.
+{{#/tab }}
+{{#tab KubeVirt}}
+
+Create a ctlptl configuration for a kind cluster with a local container registry:
```
./hack/kind-install-for-capd.sh already creates a kind cluster with a registry, the only thing it does not support is to disable CNI. 

Have you considered to improve ./hack/kind-install-for-capd.sh so we can have a consistent UX (re-use the same registry which persist across test runs, avoid ctlptl + related config file)

fabriziopandini replied:
Can we keep image registry handling as it is given that it is orthogonal to this PR?

I don't remember exact details, but it was intentional to leave the registry around. 
Probably something about avoiding to reload images e.g. in case you are testing on a fixed tag name, but I don't have time to look this up ATM

fabriziopandini commented on docs/book/src/developer/core/tilt.md:
````diff
@@ -8,26 +8,154 @@ workflow that offers easy deployments and rapid iterative builds.
+ctlptl apply -f kind-kubevirt.yaml
+```
+
+Deploy Calico as the CNI:
````
TBH I think we should have a scrpt to install CNI + MetalLB + KubeVirt, something like ./hack/kind-install-for-kubevirt.sh (that re-uses the existing script for the first part)

fabriziopandini reviewed (commented):
Thanks for writing this down.
It will be great if we can do a further step and re-use the existig scrpt for creating kind + registry as well as create a new script for automating most of the steps described in this doc

johananl commented:
@fabriziopandini I've refactored the PR and converted most of the instructions to code. We now have `make kind-cluster-kubevirt` which does almost everything. I've edited the guide accordingly.

Thanks for the suggestion :+1:

mboersma reviewed (approved):
/lgtm

The changes make sense to me. I tried this out with `make kind-cluster-kubevirt` and everything seems to work as advertised in Tilt. I did `make serve` to follow through the docs, and I think they're clear and the tab formatting works fine/

sbueringer commented on hack/kind-install-for-capd.sh:
```diff
@@ -56,21 +60,40 @@ fi
-    config_path = "/etc/containerd/certs.d"
+
+if [[ "$KIND_INFRA_PROVIDER" == "kubevirt" ]]; then
+  cat <<EOF | kind create cluster --name="$KIND_CLUSTER_NAME"  --config=-
```
Despite the name of this shell script I assume if infra provider is set to kubevirt this is not supposed to work with CAPD?
(docker mountpoint is not added)

Additionally KIND_NETWORK_IPFAMILY is not respected here

johananl replied:
>Despite the name of this shell script I assume if infra provider is set to kubevirt this is not supposed to work with CAPD?

Correct, I wanted to reuse the logic in this script with minimal changes. Do you think we should move the common logic to a new script which would be called by both `kind-install-for-capd.sh` and `kind-install-for-capk.sh`?

>Additionally KIND_NETWORK_IPFAMILY is not respected here

True. It wasn't clear to me why we might need IPv6 support and I thought IPv6 support might interfere with the logic we're adding for resolving the Docker IP prefix for the MetalLB config. I can add it if you think it's useful.

johananl replied:
Done.

sbueringer replied:
Thx

sbueringer commented:
Thx!

/lgtm

/assign @nunnatsa @fabriziopandini

chrischdi commented on Makefile:
```diff
@@ -1335,8 +1340,9 @@ clean: ## Remove generated binaries, GitBook files, Helm charts, and Tilt build
-clean-kind: ## Cleans up the kind cluster with the name $CAPI_KIND_CLUSTER_NAME
+clean-kind: ## Cleans up the kind cluster with the name $CAPI_KIND_CLUSTER_NAME and its local registry named $CAPI_KIND_REGISTRY_NAME
 	kind delete cluster --name="$(CAPI_KIND_CLUSTER_NAME)" || true
+	docker rm -f "$(CAPI_KIND_REGISTRY_NAME)"
```
I'd not like to add this. For local development it is great to preserve the registry, otherwise we always have to re-pull.

I tend to use often `make clean-kind tilt-up` for example.

johananl replied:
I see. I added this because I thought we don't want to leave a dangling container after cleanup. I can drop this line if you think it's better to keep the registry running.

chrischdi commented on hack/kind-install-for-capk.sh:
```diff
@@ -0,0 +1,105 @@
+CAPI_KIND_INFRA_PROVIDER=kubevirt "$(dirname "${BASH_SOURCE[0]}")/kind-install.sh"
+
+# Deploy Calico
+kubectl apply -f \
+  "https://raw.githubusercontent.com/projectcalico/calico/${CALICO_VERSION}/manifests/calico.yaml"
```
Question: What feature is missing in Kindnet to require calico?

johananl replied:
Quoting from the [quickstart](https://cluster-api.sigs.k8s.io/user/quick-start):

>[KubeVirt](https://kubevirt.io/) is a cloud native virtualization solution. The virtual machines we’re going to create and use for the workload cluster’s nodes, are actually running within pods in the management cluster. In order to communicate with the workload cluster’s API server, we’ll need to expose it. We are using Kind which is a limited environment. The easiest way to expose the workload cluster’s API server (a pod within a node running in a VM that is itself running within a pod in the management cluster, that is running inside a Docker container), is to use a LoadBalancer service.

>To allow using a LoadBalancer service, we can’t use the kind’s default CNI (kindnet), but we’ll need to install another CNI, like Calico.

johananl replied:
Yes, MetalLB handles the LoadBalancer services. I took these instructions almost verbatim from the official quickstart.

chrischdi replied:
> To allow using a LoadBalancer service, we can’t use the kind’s default CNI (kindnet), but we’ll need to install another CNI, like Calico.

Nit: so that doc is simply not true, because metallb is doing the LoadBalancer service job. I guess metallb does not work with kindnet and that may be the reason at the end.

Feel free to resolve here, I'm okay with that. However, might be good to clarify the docs :-)

chrischdi commented on hack/kind-install-for-capk.sh:
```diff
@@ -0,0 +1,105 @@
+CALICO_VERSION=${CAPI_CALICO_VERSION:-"v3.29.1"}
+METALLB_VERSION=${CAPI_METALLB_VERSION:-""}
+# Set manually for non-Docker runtimes - example: "172.20"
+METALLB_IP_PREFIX=${CAPI_IP_PREFIX:-""}
```
```suggestion
METALLB_IP_PREFIX=${CAPI_METALLB_IP_PREFIX:-""}
```

To match the others if we do so.

johananl replied:
Done.

chrischdi commented on hack/kind-install-for-capk.sh:
```diff
@@ -0,0 +1,105 @@
+set -o nounset
+set -o pipefail
+
+CALICO_VERSION=${CAPI_CALICO_VERSION:-"v3.29.1"}
```
Question: why do we rename the variable?

We could also do

```suggestion
CAPI_CALICO_VERSION=${CAPI_CALICO_VERSION:-"v3.29.1"}
```

To use the value given or default if not set (might be just copy/paste?)

johananl replied:
I replicated the pattern used for `CAPI_KIND_CLUSTER_NAME` to try and be consistent. I don't have strong opinions about the var names in these scripts. Do you have a preferred naming format we should use everywhere?

chrischdi replied:
just been curious. Thanks.

johananl replied:
I went through the vars again and tried to make things consistent. `CAPI_` is used as a prefix for vars settable by the user. "Internal" vars don't use this prefix.

chrischdi commented on hack/kind-install-for-capd.sh:
```diff
@@ -22,88 +22,4 @@
-    host: "localhost:${reg_port}"
-    help: "https://kind.sigs.k8s.io/docs/user/local-registry/"
-EOF
+"$(dirname "${BASH_SOURCE[0]}")/kind-install.sh"
```
Could we do something like this instead of having always new variables for different provider targets and pass in a config to kind-install.sh?:

```suggestion
KIND_NETWORK_IPFAMILY=${KIND_NETWORK_IPFAMILY:-"dual"}

CAPI_KIND_CLUSTER_CONFIG="$(cat <<EOF | kind create cluster --name="$KIND_CLUSTER_NAME"  --config=-
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
networking:
  ipFamily: ${KIND_NETWORK_IPFAMILY}
nodes:
- role: control-plane
  extraMounts:
    - hostPath: /var/run/docker.sock
      containerPath: /var/run/docker.sock
containerdConfigPatches:
- |-
  [plugins."io.containerd.grpc.v1.cri".registry]
    config_path = "/etc/containerd/certs.d"
EOF
)"

"$(dirname "${BASH_SOURCE[0]}")/kind-install.sh"
```

instead of having to pass in

chrischdi replied:
The `kind-install.sh` could then use:

```
echo "${CAPI_KIND_CLUSTER_CONFIG}" | kind create cluster --name="$KIND_CLUSTER_NAME"  --config=-
```

johananl replied:
Done.

chrischdi commented on docs/book/src/developer/core/tilt.md:
```diff
@@ -20,14 +20,64 @@ workflow that offers easy deployments and rapid iterative builds.
+
+The CAPK script needs to figure out your container runtime IP prefix to allow communication with
+workload clusters. The script assumes Docker is used. In case a different runtime is used, specify
+your container runtime's IP prefix manually (the first two octets only):
```
Maybe worth adding a short note that it get's expanded to `255.200`-`255.250` in some way?

chrischdi commented on docs/book/src/developer/core/tilt.md:
```diff
@@ -20,14 +20,70 @@ workflow that offers easy deployments and rapid iterative builds.
+{{#tabs name:"tab-management-cluster-creation" tabs:"Docker,KubeVirt"}}
+{{#tab Docker}}
+
+To create a kind cluster with CAPD, run the following:
```
```suggestion
To create a kind cluster along with a local Docker registry and the correct mounts to run CAPD, run the following:
```

chrischdi commented on docs/book/src/developer/core/tilt.md:
```diff
@@ -20,14 +20,70 @@ workflow that offers easy deployments and rapid iterative builds.
+{{#/tab }}
+{{#tab KubeVirt}}
+
+To create a kind cluster with CAPD, run the following:
```
```suggestion
To create a kind cluster with CAPK, run the following:
```

johananl replied:
Oops. Thanks.

chrischdi commented on docs/book/src/developer/core/tilt.md:
```diff
@@ -20,14 +20,70 @@ workflow that offers easy deployments and rapid iterative builds.
-A script to create a KIND cluster along with a local Docker registry and the correct mounts to run CAPD is included in the hack/ folder.
 
-To create a pre-configured cluster run:
+The following CAPI infrastructure providers are suitable for local development:
```
Any provider can be used with tilt. This doc is not only about local development.

I e.g. use the capd kind script + a different tilt-settings.yaml to develop CAPV, but also for CAPA or CAPZ in the past.

chrischdi commented on docs/book/src/developer/core/tilt.md:
```diff
@@ -36,7 +92,10 @@ kubectl cluster-info --context kind-capi-test
 ### Create a tilt-settings file
 
-Next, create a `tilt-settings.yaml` file and place it in your local copy of `cluster-api`. Here is an example that uses the components from the CAPI repo:
+Next, create a `tilt-settings.yaml` file and place it in your local copy of `cluster-api`.
```
```suggestion
Next, create a `tilt-settings.yaml` file and place it in your local copy of `cluster-api`.
Here are some examples:
```

chrischdi reviewed (commented):
Just some comments on the doc.

Great work on the bash scripts!

fabriziopandini commented on docs/book/src/developer/core/tilt.md:
```diff
@@ -20,14 +20,73 @@ workflow that offers easy deployments and rapid iterative builds.
+The following CAPI infrastructure providers are suitable for running a development environment
+without using real machines or cloud resources:
+
+- [CAPD](https://github.com/kubernetes-sigs/cluster-api/blob/main/test/infrastructure/docker/README.md) - uses Docker containers as workload cluster nodes
```
nit: it also supports in memory machines now, see https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20250124-From%20CAPD(docker)%20to%20CAPD(dev)%20.md

johananl replied:
I changed the phrasing.

fabriziopandini commented on hack/kind-install-for-capk.sh:
```diff
@@ -0,0 +1,126 @@
+done
+echo "Pod created!"
+
+echo "Waiting for speaker pod to be created..."
```
```suggestion
echo "Waiting for MetalLB speaker pod to be created..."
```

fabriziopandini commented on hack/kind-install-for-capk.sh:
```diff
@@ -0,0 +1,126 @@
+  -l app=metallb,component=speaker -o jsonpath="{.items[0].metadata.name}" 2>/dev/null) ]]; do
+  sleep 2
+done
+echo "Pod created!"
```
```suggestion
echo "MetalLB speaker Pod created!"
```

fabriziopandini commented on hack/kind-install-for-capk.sh:
```diff
@@ -0,0 +1,126 @@
+done
+echo "Pod created!"
+
+echo "Waiting for pods to become ready..."
```
```suggestion
echo "Waiting for MetalLB pods to become ready..."
```

fabriziopandini commented on hack/kind-install-for-capk.sh:
```diff
@@ -0,0 +1,126 @@
+  -l app=metallb,component=controller --for=condition=Ready --timeout=10m
+kubectl wait pods -n metallb-system \
+  -l app=metallb,component=speaker --for=condition=Ready --timeout=2m
+echo "Pods ready!"
```
```suggestion
echo "MetalLB Pods ready!"
```

fabriziopandini reviewed (commented):
just few nits otherwise lgtm, but I'm not sure when I will have bandwidth to test, hopefully someone could get there before me

dmvolod commented on hack/kind-install-for-capk.sh:
```diff
@@ -0,0 +1,126 @@
+  "https://raw.githubusercontent.com/metallb/metallb/${METALLB_VERSION}/config/manifests/metallb-native.yaml"
+
+echo "Waiting for controller pod to be created..."
+while [[ -z $(kubectl -n metallb-system get pods \
+  -l app=metallb,component=controller -o jsonpath="{.items[0].metadata.name}" 2>/dev/null) ]]; do
+  sleep 2
+done
```
It seems to waiting for MetalLB Deployment and DaemonSet pods can be a bit simplified with `kubectl wait` option. 

```
kubectl wait --for condition=available deployment controller  -n metallb-system --timeout 150s
kubectl wait --for condition=ready pod -l app=metallb,component=speaker -n metallb-system --timeout 150s
```

johananl replied:
I used the loop because I kept getting errors because the script was trying to check for `Ready` status on a non-existent pod.

Good point regarding checking for `available` status on the **deployment** rather than the pods. This works. But looks like we don't have an equivalent for daemon sets.

We could do what you suggested and check for `available` deployment and `ready` pod, however there will be a race between speaker pod creation and the wait on speaker pod.

johananl replied:
I've adopted your suggestion for now to make progress. Maybe we can leave the race be since it's probably quite rare that it would trip someone.

dmvolod replied:
Yes, totally agree about potential race between controller and speaker pods.

Probably, I've found solution for the DaemonSet wait for ready condition:
```
kubectl wait -n metallb-system daemonsets.apps speaker --for=jsonpath='{.status.numberReady}'=$(kubectl get daemonsets.apps speaker -n metallb-system  -ojsonpath='{.status.currentNumberScheduled}'
```

johananl replied:
Thanks, but it doesn't work: When trying to run

```
kubectl wait pods -n metallb-system -l app=metallb,component=speaker --for condition=Ready --timeout 5m
```

right after

```
kubectl wait -n metallb-system daemonsets.apps speaker \
  --for jsonpath='{.status.numberReady}'="$(kubectl get daemonsets.apps speaker -n metallb-system -ojsonpath='{.status.currentNumberScheduled}')" --timeout 5m
```

```
error: no matching resources found
```

dmvolod replied:
It's strange, but let's keep the original approach. Thank you.

johananl replied:
Actually I've already adopted your suggestion partially, just without waiting for the daemonset (see current state of code). Think we should go back to the `sleep` loops?

dmvolod replied:
No, adopted part is good. Without waiting for the DaemonSet approach, as it's seems to not working in some cases.

Probably this code should work. I tried to run it locally and it's working for me
```
kubectl wait -n metallb-system daemonsets.apps -l app=metallb,component=speaker --for=jsonpath='{.status.numberReady}'=$(kubectl get daemonsets.apps -l app=metallb,component=speaker -n metallb-system  -ojsonpath='{.items[0].status.currentNumberScheduled}')
```

johananl replied:
It's "working" for me only because by the time the controller deployment has converged the speaker DS has converged, too. I ran

```
kubectl wait -n metallb-system daemonsets.apps -l app=metallb,component=speaker \
  --for=jsonpath='{.status.numberReady}'="$(kubectl get daemonsets.apps -l app=metallb,component=speaker -n metallb-system  -ojsonpath='{.items[0].status.currentNumberScheduled}')"

kubectl wait pods -n metallb-system -l app=metallb,component=speaker --for condition=Ready --timeout 5m
```

for testing and got:

```
error: no matching resources found
```

johananl replied:
Let's keep the current state and if anyone ever complains about the race we switch to `sleep`.

johananl commented:
I'd like to merge this ASAP since we depend on this PR for the Node Bootstrapping working group (https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/community/20241112-node-bootstrapping.md).

I think we can follow up with fixes or enhancements at any point later on but for now let's have something that works. Also, this is a dev tool rather than a user-facing feature so IMO it's OK if it's not 100% polished.

chrischdi commented on hack/kind-install.sh:
```diff
@@ -0,0 +1,90 @@
+set -o nounset
+set -o pipefail
+
+if [[ "${TRACE-0}" == "1" ]]; then
+    set -o xtrace
+fi
```
Last nit, would be good to have this also in the other two shell scripts :-)

chrischdi reviewed (commented):
Last nit, happy to merge afterwards :-)

johananl commented:
Thanks @chrischdi. I think we should be good to go.

sbueringer commented:
/assign

chrischdi commented:
lgtm from my side, pending stefans review :-)

sbueringer commented:
Thank you!

/lgtm
/approve
